File: day09/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = true;
// const DEBUG = false;

const text = await Deno.readTextFile('input.txt');
if (DEBUG) console.debug('text', text);

const FREE_SPACE = Infinity; //possible deno bug - array of length>=100 containing value -1 .indexOf(-1) returns -1

const diskMap = text
  .split('')
  .filter((char) => char.trim() != '')
  .map((char) => parseInt(char))
  .map((length, ix) => {
    return new Array(length).fill(ix % 2 ? FREE_SPACE : Math.round(ix / 2));
  })
  .flat();
if (DEBUG) console.debug('diskMap', diskMap);

const part1 = (diskMap: number[]) => {
  let packedEndIx = -1;
  for (let fromIx = diskMap.length - 1; fromIx >= 0; fromIx--) {
    const fromFile = diskMap[fromIx];
    if (fromFile == FREE_SPACE) continue;
    const toIx = diskMap.indexOf(FREE_SPACE, packedEndIx + 1);
    if (DEBUG) console.debug({ fromIx, fromFile, toIx });
    if (toIx == FREE_SPACE || toIx > fromIx) break;
    diskMap[toIx] = fromFile;
    diskMap[fromIx] = FREE_SPACE;
    packedEndIx = toIx;
  }
  if (DEBUG) console.debug('diskMap', diskMap);

  const checksum = diskMap.reduce((acc, fileId, ix) => {
    return acc + (fileId == FREE_SPACE ? 0 : fileId * ix);
  }, 0);
  console.log('part 1', checksum);
};

const part2 = (diskMap: number[]) => {
  for (let fromFile = Math.max(...diskMap.filter((i) => i != FREE_SPACE)); fromFile >= 0; fromFile--) {
    const fromIx = diskMap.indexOf(fromFile);
    const fromLength = diskMap.lastIndexOf(fromFile) - fromIx + 1;

    let toLength = 0;
    let toIx = FREE_SPACE;
    for (let testIx = diskMap.indexOf(FREE_SPACE); testIx < fromIx; testIx++) {
      switch (diskMap[testIx]) {
        case FREE_SPACE:
          toLength++;
          if (toIx == FREE_SPACE) toIx = testIx;
          break;
        default:
          toLength = 0;
          toIx = FREE_SPACE;
      }
      if (toLength == fromLength) {
        diskMap.fill(fromFile, toIx, toIx + fromLength);
        diskMap.fill(FREE_SPACE, fromIx, fromIx + fromLength);
        if (DEBUG) console.debug({ fromFile, fromIx, fromLength, toIx });
        break;
      }
    }
  }
  if (DEBUG) console.debug('diskMap', diskMap);
  const checksum = diskMap.reduce((acc, fileId, ix) => {
    return acc + (fileId == FREE_SPACE ? 0 : fileId * ix);
  }, 0);
  console.log('part 2', checksum);
};

// part1([...diskMap]);
part2([...diskMap]);

File: day09/solution.go
//go:build ignore

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	data := readData("input.txt")
	// fmt.Println(data)
	part1(copyDiskmap(data))
	// fmt.Println(data)
	part2(copyDiskmap(data))
}

func part2(diskMap []int) {
	lastFileId := -1
	for fromIxEnd := len(diskMap) - 1; fromIxEnd >= 0; fromIxEnd-- {
		fileId := diskMap[fromIxEnd]
		if fileId == -1 || fileId == lastFileId {
			continue
		}
		lastFileId = fileId
		fromIxStart := indexOf(diskMap, fileId)
		fileLength := fromIxEnd - fromIxStart + 1
		// fmt.Println("fileId",fileId,"fromIxStart",fromIxStart,"fromIxEnd",fromIxEnd,"fileLength",fileLength)
		spaceLength := 0
		spaceStartIx := -1
		for spaceScanIx := 0; spaceScanIx < fromIxStart; spaceScanIx++ {
			if diskMap[spaceScanIx] == -1 {
				spaceLength++
				if spaceStartIx == -1 {
					spaceStartIx = spaceScanIx
				}
				if spaceLength == fileLength {
					//BUG: we're insertint 1 index higher than actual
					for i := 0; i < fileLength; i++ {
						diskMap[spaceStartIx+i] = fileId
						diskMap[fromIxStart+i] = -1
					}
					// fmt.Println("move file",fileId,"from",fromIxStart,"length",fileLength,"to",spaceStartIx)
					break
				}
			} else {
				spaceLength = 0
				spaceStartIx = -1
			}
		}
	}
	checksum := 0
	for i, fileId := range diskMap {
		if fileId == -1 {
			continue
		}
		checksum += i * fileId
	}
	// fmt.Println(diskMap)
	fmt.Println("part 2:", checksum)
}

func indexOf(items []int, search int) int {
	for ix, item := range items {
		if item == search {
			return ix
		}
	}
	return -1
}

func part1(diskMap []int) {
	lastToIx := 0
	for fromIx := len(diskMap) - 1; fromIx >= 0; fromIx-- {
		fileId := diskMap[fromIx]
		if fileId == -1 {
			continue
		}
		for toIx := lastToIx; toIx < fromIx; toIx++ {
			if diskMap[toIx] == -1 {
				// fmt.Println("move file",fileId,"from",fromIx,"to",toIx)
				diskMap[toIx] = fileId
				diskMap[fromIx] = -1
				lastToIx = toIx
				break
			}
		}
	}
	// fmt.Println(diskMap)
	checksum := 0
	for i, fileId := range diskMap {
		if fileId == -1 {
			continue
		}
		checksum += i * fileId
	}
	fmt.Println("part 1:", checksum)
}

func copyDiskmap(data []int) []int {
	newData := make([]int, len(data))
	copy(newData, data)
	return newData
}

func readData(filename string) []int {
	f, _ := os.Open(filename)
	defer f.Close()
	var data []int
	scanner := bufio.NewScanner(f)
	scanner.Split(bufio.ScanRunes)
	charIx := 0
	for scanner.Scan() {
		length, _ := strconv.Atoi(scanner.Text())
		fileId := -1
		if charIx%2 == 0 {
			fileId = charIx / 2
		}
		fileData := make([]int, length)
		for i := range fileData {
			fileData[i] = fileId
		}
		data = append(data, fileData...)
		charIx++
	}
	return data
}
